Compiler panic on type inference for impl Generator yielding a reference

I'm getting

> error: internal compiler error: librustc/infer/error_reporting/mod.rs:184: ReEmpty
> thread 'main' panicked at 'Box<Any>', librustc_errors/lib.rs:554:9

when trying to compile
```rust
#![feature(generators, generator_trait)]
use std::ops::Generator;
fn example() -> impl Generator {
    || { yield &1 }
}
```
on rustc 1.29.0-nightly (e5f6498d3 2018-07-10) - [Playground demo](https://play.rust-lang.org/?gist=2421dbda0b8b259b3efec9897a58f2d6&version=nightly&mode=debug&edition=2015).

This seems to happen when I `yield` a reference type and don't specify the exact type in `impl Generator<Yield=&Whatever>`.

We don't generate an error message for the `ReEmpty` region here and panic instead. Here is the message that should be printed with a placeholder:
```
error[E0700]: hidden type for `impl Trait` captures lifetime that does not appear in bounds
 --> test-region.rs:3:17
  |
3 | fn example() -> impl Generator {
  |                 ^^^^^^^^^^^^^^
  |
  = note: hidden type `[generator@test-region.rs:4:5: 4:20 for<'r> {i32, &'r i32, ()}]` captures BUG[ReEmpty]
```
Should `ReEmpty` be allowed for impl Trait? cc @rust-lang/compiler

--- a/src/test/ui/generator/issue-52304.rs
+++ b/src/test/ui/generator/issue-52304.rs
@@ -1,0 +1,11 @@
+// check-pass
+
+#![feature(generators, generator_trait)]
+
+use std::ops::Generator;
+
+pub fn example() -> impl Generator {
+    || yield &1
+}
+
+fn main() {}

